network/p2p/topic_validator_test.go

```diff
@@ -109,31 +107,349 @@ func TestTopicValidator(t *testing.T) {
+		}
+
+		return ids.ByNodeID(fid)
+	})
+
+	// node1 is connected to node2, and the an1 is connected to node1
+	// an1 <-> sn1 <-> sn2
+	require.NoError(t, sn1.AddPeer(context.TODO(), *host.InfoFromHost(sn2.host)))
+	require.NoError(t, an1.AddPeer(context.TODO(), *host.InfoFromHost(sn1.host)))
+
+	// sn1 and sn2 subscribe to the topic with the topic validator
+	sub1, err := sn1.Subscribe(topic, authorizedSenderValidator)
+	require.NoError(t, err)
+	sub2, err := sn2.Subscribe(topic, authorizedSenderValidator)
+	require.NoError(t, err)
+	sub3, err := an1.Subscribe(topic)
+	require.NoError(t, err)
+
+	// assert that the nodes are connected as expected
+	require.Eventually(t, func() bool {
+		return len(sn1.pubSub.ListPeers(topic.String())) > 0 &&
+			len(sn2.pubSub.ListPeers(topic.String())) > 0 &&
+			len(an1.pubSub.ListPeers(topic.String())) > 0
+	}, 3*time.Second, 100*time.Millisecond)
+
+	timedCtx, cancel5s := context.WithTimeout(context.Background(), 1000*time.Second)
+	defer cancel5s()
+	// create a dummy block proposal to publish from our SN node
+	header := unittest.BlockHeaderFixture()
+	data1 := getMsgFixtureBz(t, &messages.BlockProposal{Header: &header})
+
+	// sn2 publishes the block proposal, sn1 and an1 should receive the message because
+	// SN nodes are authorized to send block proposals
+	err = sn2.Publish(timedCtx, topic, data1)
+	require.NoError(t, err)
+
+	// sn1 gets the message
+	checkReceive(timedCtx, t, data1, sub1, nil, true)
+
+	// sn2 also gets the message (as part of the libp2p loopback of published topic messages)
+	checkReceive(timedCtx, t, data1, sub2, nil, true)
+
+	// an1 also gets the message
+	checkReceive(timedCtx, t, data1, sub3, nil, true)
+
+	timedCtx, cancel2s := context.WithTimeout(context.Background(), 2*time.Second)
+	defer cancel2s()
+	header = unittest.BlockHeaderFixture()
+	data2 := getMsgFixtureBz(t, &messages.BlockProposal{Header: &header})
+
+	// the access node now publishes the block proposal message, AN are not authorized to publish block proposals
+	// the message should be rejected by the topic validator on sn1
+	err = an1.Publish(timedCtx, topic, data2)
+	require.NoError(t, err)
+
+	// an1 receives its own message
+	checkReceive(timedCtx, t, data2, sub3, nil, true)
+
+	var wg sync.WaitGroup
+
+	// sn1 does NOT receive the message due to the topic validator
+	timedCtx, cancel := context.WithTimeout(context.Background(), time.Second)
+	defer cancel()
+	checkReceive(timedCtx, t, nil, sub1, &wg, false)
+
+	// sn2 also does not receive the message via gossip from the sn1 (event after the 1 second hearbeat)
+	timedCtx, cancel = context.WithTimeout(context.Background(), 2*time.Second)
+	defer cancel()
+	checkReceive(timedCtx, t, nil, sub2, &wg, false)
+
+	unittest.RequireReturnsBefore(t, wg.Wait, 5*time.Second, "could not receive message on time")
+
+	// expecting 1 warn calls for each rejected message from unauthorized node
+	require.Equalf(t, 1, hookCalls, "expected 1 warning to be logged")
```

Comment:
Shouldn't we get at least 2 warnings? (One from each consensus node which rejected the block proposal from AN)

Or does SN2 never receive the proposal message at all? (it is discarded before being gossiped by SN1)

Reply:
Yes the second point is true that SN2 does not receive the message via gossip, this is noted in the comments of the test as well